lte/gateway/c/oai/tasks/amf/amf_app_ue_context_and_proc.h

```diff
@@ -0,0 +1,1001 @@
+      const identity_type2_t type, success_cb_t success, failure_cb_t failure);
+  int amf_proc_identification_complete(
+      const amf_ue_ngap_id_t ue_id, imsi_t* const imsi, imei_t* const imei,
+      imeisv_t* const imeisv, uint32_t* const tmsi);
+};
+
+class nas_amf_auth_proc_t {
+ public:
+  nas_amf_common_proc_t amf_com_proc;
+  nas5g_timer_t T3560; /* Authentication timer         */
+#define AUTHENTICATION_COUNTER_MAX 5
+  unsigned int retransmission_count;
+#define AMF_AUTHENTICATION_SYNC_FAILURE_MAX 2
+  unsigned int
+      sync_fail_count; /* counter of successive AUTHENTICATION FAILURE messages
+                      from the UE with AMF cause #21 "synch failure" */
+  unsigned int mac_fail_count;
+  amf_ue_ngap_id_t ue_id;
+  bool is_cause_is_registered;  //  could also be done by seeking parent
+                                //  procedure
+  ksi_t ksi;
+  uint8_t rand[AUTH_RAND_SIZE]; /* Random challenge number  */
+  uint8_t autn[AUTH_AUTN_SIZE]; /* Authentication token     */
+  imsi_t* unchecked_imsi;
+  int amf_cause;
+};
+
+typedef struct nas5g_cn_proc_s {
+  nas5g_base_proc_t base_proc;
+  cn5g_proc_type_t type;
+} nas5g_cn_proc_t;
+
+typedef struct nas5g_cn_procedure_s {
+  nas5g_cn_proc_t* proc;
+  LIST_ENTRY(nas5g_cn_procedure_s) entries;
+} nas5g_cn_procedure_t;
+
+class nas_5g_auth_info_proc_t {
+ public:
+  nas5g_cn_proc_t cn_proc;
+  success_cb_t success_notif;
+  failure_cb_t failure_notif;
+  bool request_sent;
+  uint8_t nb_vectors;
+  int nas_cause;
+  amf_ue_ngap_id_t ue_id;
+  bool resync;  // Indicates whether the authentication information is requested
+                // due to sync failure
+};
+
+nas_5g_auth_info_proc_t* nas_new_5gcn_auth_info_procedure(
+    amf_context_t* const amf_context);
+
+class m5g_authentication : public amf_context_t {
+ public:
+  int amf_proc_authentication_ksi(
+      amf_context_t* amf_context,
+      nas_amf_specific_proc_t* const amf_specific_proc, ksi_t ksi,
+      const uint8_t* const rand, const uint8_t* const autn,
+      success_cb_t success, failure_cb_t failure);
+  int amf_proc_authentication(
+      amf_context_t* amf_context,
+      nas_amf_specific_proc_t* const amf_specific_proc, success_cb_t success,
+      failure_cb_t failure);
+  int amf_proc_authentication_failure(
+      amf_ue_ngap_id_t ue_id, int amf_cause, const_bstring auts);
+  int amf_proc_authentication_complete(
+      amf_ue_ngap_id_t ue_id, AuthenticationResponseMsg* msg, int amf_cause,
+      const unsigned char* res);
+  int amf_registration_security(amf_context_t* amf_context);
+  void set_notif_callbacks_for_5g_auth_proc(nas_amf_auth_proc_t* auth_proc);
+  void set_callbacks_for_5g_auth_proc(nas_amf_auth_proc_t* auth_proc);
+  void set_callbacks_for_5g_auth_info_proc(
+      nas_5g_auth_info_proc_t* auth_info_proc);
+  int amf_send_authentication_request(
+      amf_context_t* amf_context, nas_amf_auth_proc_t* auth_proc);
+};
```

Comment:
does not make much sense to me, why inherit amf_context? why group them in a class? just place all these as utility functions in a fiveg authentication procedure header.

Reply:
fixed